File: langs/python/stdlib.py
#!/usr/bin/env python
import enum
import math


class Direction(enum.Enum):
    North = "North"
    South = "South"
    East = "East"
    West = "West"

    @property
    def opposite(self):
        return {
            Direction.East: Direction.West,
            Direction.West: Direction.East,
            Direction.South: Direction.North,
            Direction.North: Direction.South,
        }[self]

    @property
    def to_coords(self):
        return {
            Direction.East: Coords(1, 0),
            Direction.West: Coords(-1, 0),
            Direction.South: Coords(0, 1),
            Direction.North: Coords(0, -1),
        }[self]


class ActionType(enum.Enum):
    Attack = "Attack"
    Move = "Move"


class Coords(tuple):
    def __new__(cls, x, y):
        self = super().__new__(cls, [x, y])
        return self

    def __repr__(self):
        return "({self.x}, {self.y})"

    @property
    def x(self):
        return self[0]

    @property
    def y(self):
        return self[1]

    def distance(self, other):
        return math.sqrt((other.x - self.x) ** 2 + (other.y - self.y) ** 2)

    def walking_distance(self, other):
        return abs(other.x - self.x) + abs(other.y - self.y)

    def coords_around(self):
        [self + direction for direction in Direction]

    def towards(self, other):
        diff = other - self
        angle = math.atan2(diff.y, diff.x)
        if abs(angle) > math.pi / 4:
            if diff.y > 0:
                return Direction.North
            else:
                return Direction.South
        else:
            if diff.x > 0:
                return Direction.East
            else:
                return Direction.West

    def __add__(self, other):
        return Coords(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        return Coords(self.x - other.x, self.y - other.y)

    def __mul__(self, n):
        return Coords(self.x * n, self.y * n)


class Team(enum.Enum):
    Red = "Red"
    Blue = "Blue"

    @property
    def opposite(self):
        if self == Team.Red:
            return Team.Blue
        else:
            return Team.Red


class ObjType(enum.Enum):
    Unit = "Unit"
    Terrain = "Terrain"


class Obj:
    def __init__(self, obj):
        self.__data = obj

    @property
    def coords(self):
        return Coords(*self.__data["coords"])

    @property
    def id(self):
        return self.__data["id"]

    @property
    def obj_type(self):
        return ObjType(self.__data["obj_type"])

    @property
    def team(self):
        if self.obj_type == ObjType.Unit:
            return Team(self.__data["team"])

    @property
    def health(self):
        if self.obj_type == ObjType.Unit:
            return self.__data["health"]


class State:
    def __init__(self, state):
        self.__data = state

    @property
    def turn_num(self):
        return self.__data["turn_num"]

    @property
    def our_team(self):
        return Team(self.__data["team"])

    @property
    def other_team(self):
        return self.our_team.opposite()

    def ids_by_team(self, team):
        return self.__data["teams"][team.value]

    def obj_by_id(self, id):
        return Obj(self.__data["objs"][id])

    def objs_by_team(self, team):
        return [self.obj_by_id(id) for id in self.ids_by_team(team)]

    def id_by_coords(self, coords):
        return self.__data["grid"][coords.x][coords.y]

    def obj_by_coords(self, coords):
        return self.obj_by_id(self.id_by_coords(coords))


class Action:
    def __init__(self, type, direction):
        self.type = type
        self.direction = direction

    def __repr__(self):
        return f"<Action: {self.type} {self.direction}>"

    @staticmethod
    def move(direction):
        return Action(ActionType.Move, direction)

    @staticmethod
    def attack(direction):
        return Action(ActionType.Attack, direction)


def __format_err(exc):
    loc = None
    tb = exc.__traceback__
    while tb:
        if tb.tb_frame.f_code.co_filename == "<robot>":
            loc = {
                "start": (tb.tb_lineno, None),
                "end": None,
            }
        tb = tb.tb_next
    return {
        "message": str(exc),
        "loc": loc,
    }


def __main(state, scope=globals()):
    def __validate_function(name, argcount, mandatory):
        f = scope.get(name)
        if not callable(f):
            if mandatory:
                raise TypeError(f"You must define a '{name}' function")
        else:
            if f.__code__.co_argcount != argcount:
                raise TypeError(
                    f"Your {name} function must accept {argcount} arguments"
                )
        return f

    import sys, io

    had_stdout, old_stdout = (True, sys.stdout) if hasattr(sys, "stdout") else (False, None)
    logbuf = sys.stdout = io.StringIO()

    state = State(state)
    try:
        _robot = __validate_function("_robot", 3, True)
        _init_turn = __validate_function("_init_turn", 1, False)
    except Exception as e:
        return {"robot_outputs": {"Err": {"InitError": __format_err(e)}}}

    if callable(_init_turn):
        _init_turn(state)

    robot_outputs = {}
    for id in state.ids_by_team(state.our_team):
        debug_table = {}

        def debug(key, val):
            debug_table[key] = str(val)

        try:
            action = _robot(state, state.obj_by_id(id), debug)
            if not isinstance(action, Action):
                raise TypeError("Your _robot function must return an Action")
        except Exception as e:
            result = {"Err": __format_err(e)}
        else:
            result = {
                "Ok": {"type": action.type.value, "direction": action.direction.value}
            }
        robot_outputs[id] = {"action": result, "debug_table": debug_table}

    if had_stdout:
        sys.stdout = old_stdout
    else:
        del sys.stdout

    logbuf.seek(0)
    logs = logbuf.readlines()
    logbuf.close()

    return {"robot_outputs": {"Ok": robot_outputs}, "logs": logs}


del enum

if __name__ == '__main__':
    __builtins__.__dict__.update(globals())
    import sys, json, runpy

    module = sys.argv[1]
    module = runpy.run_path(module)
    print('__rr_init:{"Ok":null}', flush=True)
    for inp in sys.stdin:
        inp = json.loads(inp)
        output = __main(inp, scope=module)
        sys.stdout.write("__rr_output:")
        json.dump(output, sys.stdout)
        print(flush=True)

File: langs/javascript/stdlib.js
// "use strict";

// https://2ality.com/2020/01/enum-pattern.html
// https://github.com/rauschma/enumify/blob/master/ts/src/index.ts
class Enum {
  static closeEnum () {
    const enumValues = []

    for (const [name, staticInstance] of Object.entries(this)) {
      staticInstance.value = name
      staticInstance.ordinal = enumValues.length
      enumValues.push(staticInstance)
    }

    // Important: only add more static properties *after* processing the enum entries
    this.enumValues = enumValues
  }

  static valueOf (str) {
    return this.enumValues.find(val => val.enumKey === str)
  }

  // INSTANCE

  toString () {
    return `${this.constructor.name}.${this.enumKey}`
  }
}

class Direction extends Enum {
  static East = new Direction()
  static West = new Direction()
  static South = new Direction()
  static North = new Direction()
  static _ = this.closeEnum()

  get opposite () {
    switch (this) {
      case Direction.East:
        return Direction.West
      case Direction.West:
        return Direction.East
      case Direction.North:
        return Direction.South
      case Direction.South:
        return Direction.North
    }
  }

  get toCoords () {
    switch (this) {
      case Direction.East:
        return new Coords(1, 0)
      case Direction.West:
        return new Coords(-1, 0)
      case Direction.North:
        return new Coords(0, 1)
      case Direction.South:
        return new Coords(0, -1)
    }
  }
}

class Coords {
  constructor (x, y) {
    this.x = x
    this.y = y
  }

  toString () {
    return `(${this.x}, ${this.y})`
  }

  distance (other) {
    return Math.sqrt((other.x - this.x) ** 2 + (other.y - this.y) ** 2)
  }

  walkingDistance (other) {
    return Math.abs(other.x - this.x) + Math.abs(other.y - this.y)
  }

  towards (other) {
    const diff = this.sub(other)
    const angle = Math.atan2(diff.y, diff.x)
    if (Math.abs(angle) > Math.PI / 4) {
      if (diff.y > 0) return Direction.North
      else return Direction.South
    } else {
      if (diff.x > 0) return Direction.East
      else return Direction.West
    }
  }

  add (other) {
    return new Coords(this.x + other.x, this.y + other.y)
  }

  sub (other) {
    return new Coords(this.x - other.x, this.y - other.y)
  }

  mul (n) {
    return new Coords(this.x * n, this.y * n)
  }
}

class Team extends Enum {
  static Red = new Team()
  static Blue = new Team()
  static _ = this.closeEnum()

  get opposite () {
    if (this === Team.Red) {
      return Team.Blue
    } else return Team.Red
  }
}

class ObjType extends Enum {
  static Unit = new ObjType()
  static Terrain = new ObjType()
  static _ = this.closeEnum()
}

class Obj {
  constructor (obj) {
    this.__data = obj
  }

  get coords () {
    return new Coords(...this.__data['coords'])
  }

  get id () {
    return this.__data['id']
  }

  get objType () {
    return ObjType.valueOf(this.__data['obj_type'])
  }

  get team () {
    if (this.objType === ObjType.Unit) {
      return Team.valueOf(this.__data['team'])
    }
  }

  get health () {
    if (this.objType === ObjType.Unit) {
      return this.__data['health']
    }
  }
}

class State {
  constructor (state) {
    this.__data = state
  }

  get turnNum () {
    return this.__data['turnNum']
  }

  get ourTeam () {
    return Team.valueOf(this.__data['team'])
  }

  get otherTeam () {
    return this.ourTeam.opposite()
  }

  idsByTeam (team) {
    return this.__data['teams'][team.value]
  }

  objById (id) {
    return new Obj(this.__data['objs'][id])
  }

  objsByTeam (team) {
    return this.idsByTeam(team).map(this.objById)
  }

  idByCoords (coords) {
    return this.__data['grid'][coords.x][coords.y]
  }

  objByCoords (coords) {
    return this.objById(this.idByCoords(coords))
  }
}

class ActionType extends Enum {
  static Attack = new ActionType()
  static Move = new ActionType()
  static _ = this.closeEnum()
}

class Action {
  constructor (type, direction) {
    this.type = type
    this.direction = direction
  }

  toString () {
    return `<Action: ${this.type} ${this.direction}>`
  }

  static move (direction) {
    return new Action(ActionType.Move, direction)
  }

  static attack (direction) {
    return new Action(ActionType.Attack, direction)
  }
}

function __format_err (err, incl_err = false, init_err = false) {
  const e = {
    start: [0, 0],
    end: [0, 0],
    message: err.toString(),
  }
  return incl_err ? { Err: init_err ? { InitError: e } : e } : e
}


function __main (stateData) {
  function __validateFunction(f, argcount, mandatory) {
    if (typeof f !== "function") {
      if (mandatory) {
        throw new TypeError(`You must define a '${name}' function`)
      }
    } else {
      if (f.length !== argcount) {
        throw new TypeError(
          `Your ${name} function must accept ${argcount} arguments`
        )
      }
    }
  }

  const state = new State(stateData)

  try {
    __validateFunction(globalThis._robot, 3, true)
    __validateFunction(globalThis._initTurn, 1, true)
  } catch (e) {
    return {"robot_outputs": {"Err": {"InitError": __format_err(e)}}}
  }

  if (typeof globalThis._initTurn === 'function') {
    globalThis._initTurn(state)
  }

  const robotOutputs = {}
  for (const id of state.idsByTeam(state.ourTeam)) {
    const debug_table = {}

    const debug = (key, val) => {
      debug_table[key] = val
    }

    let result
    try {
      result = { Ok: globalThis._robot(stateData, state.objById(id), debug) }
    } catch (e) {
      result = { Err: __format_err(e) }
    }
    robotOutputs[id] = { action: result, debug_table }
  }
  return { robot_outputs: { Ok: robotOutputs }, logs: [] }
}
